pkg/remote/gitlab/gitlab.go

```diff
@@ -0,0 +1,374 @@
+	URL         string
+	Client      string
+	Secret      string
+	AllowedOrgs []string
+	Open        bool
+	PrivateMode bool
+	SkipVerify  bool
+
+	cache *lru.Cache
+}
+
+func init() {
+	remote.Register("gitlab", NewDriver)
+}
+
+func NewDriver(conf *config.Config) (remote.Remote, error) {
+	var gitlab = Gitlab{
+		URL:         conf.Gitlab.URL,
+		Client:      conf.Gitlab.Client,
+		Secret:      conf.Gitlab.Secret,
+		AllowedOrgs: conf.Gitlab.Orgs,
+		Open:        conf.Gitlab.Open,
+		SkipVerify:  conf.Gitlab.SkipVerify,
+	}
+	var err error
+	gitlab.cache, err = lru.New(1028)
+	if err != nil {
+		return nil, err
+	}
+
+	// the URL must NOT have a trailing slash
+	if strings.HasSuffix(gitlab.URL, "/") {
+		gitlab.URL = gitlab.URL[:len(gitlab.URL)-1]
+	}
+	return &gitlab, nil
+}
+
+func (r *Gitlab) Login(token, secret string) (*common.User, error) {
+	client := NewClient(r.URL, token, r.SkipVerify)
+	var login, err = client.CurrentUser()
+	if err != nil {
+		return nil, err
+	}
+	user := common.User{}
+	user.Login = login.Username
+	user.Email = login.Email
+	user.Token = token
+	user.Secret = secret
+	return &user, nil
+}
+
+// Orgs fetches the organizations for the given user.
+func (r *Gitlab) Orgs(u *common.User) ([]string, error) {
+	return nil, nil
+}
+
+// Repo fetches the named repository from the remote system.
+func (r *Gitlab) Repo(u *common.User, owner, name string) (*common.Repo, error) {
+	client := NewClient(r.URL, u.Token, r.SkipVerify)
+	id := ns(owner, name)
+	repo_, err := client.Project(id)
+	if err != nil {
+		return nil, err
+	}
+
+	repo := &common.Repo{}
+	repo.Owner = owner
+	repo.Name = name
+	repo.FullName = repo_.PathWithNamespace
+	repo.Link = repo_.Url
+	repo.Clone = repo_.HttpRepoUrl
+	repo.Branch = "master"
+
+	if repo_.DefaultBranch != "" {
+		repo.Branch = repo_.DefaultBranch
+	}
+
+	if r.PrivateMode {
+		repo.Private = true
+	}
+	return repo, err
+}
+
+// Perm fetches the named repository from the remote system.
+func (r *Gitlab) Perm(u *common.User, owner, name string) (*common.Perm, error) {
+	key := fmt.Sprintf("%s/%s/%s", u.Login, owner, name)
+	val, ok := r.cache.Get(key)
+	if ok {
+		return val.(*common.Perm), nil
+	}
+
+	client := NewClient(r.URL, u.Token, r.SkipVerify)
+	id := ns(owner, name)
+	repo, err := client.Project(id)
+	if err != nil {
+		return nil, err
+	}
+	m := &common.Perm{}
+	m.Admin = IsAdmin(repo)
+	m.Pull = IsRead(repo)
+	m.Push = IsWrite(repo)
+	r.cache.Add(key, m)
+	return m, nil
+}
+
+// GetScript fetches the build script (.drone.yml) from the remote
+// repository and returns in string format.
+func (r *Gitlab) Script(user *common.User, repo *common.Repo, build *common.Build) ([]byte, error) {
+	var client = NewClient(r.URL, user.Token, r.SkipVerify)
+	var path = ns(repo.Owner, repo.Name)
+	return client.RepoRawFile(path, build.Commit.Sha, ".drone.yml")
+}
+
+func (r *Gitlab) Status(u *common.User, repo *common.Repo, b *common.Build) error {
+	// how can I get Gitlab status
```

Comment:
Currently gitlab doesn't support status for commits and events, also if we want get MR status in gitlab we need implement a special plugin for gitlab, gitlab uses API to fetch build status on client side. But for now we skip this.